src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -954,74 +922,69 @@ MulticopterPositionControl::control_manual(float dt)
-		req_vel_sp_xy(1) = math::expo_deadzone(_manual.y, _params.xy_vel_man_expo, _params.hold_xy_dz);
+		man_vel_sp(0) = math::expo_deadzone(_manual.x, _params.xy_vel_man_expo, _params.hold_dz);
+		man_vel_sp(1) = math::expo_deadzone(_manual.y, _params.xy_vel_man_expo, _params.hold_dz);
 
-		/* saturarate such that magnitude is never larger than 1 */
-		if (req_vel_sp_xy.length() > 1.0f) {
-			req_vel_sp_xy = req_vel_sp_xy.normalized();
+		const float man_vel_hor_length = ((matrix::Vector2f)man_vel_sp.slice<2, 1>(0, 0)).length();
+
+		/* saturate such that magnitude is never larger than 1 */
+		if (man_vel_hor_length > 1.0f) {
+			man_vel_sp(0) /= man_vel_hor_length;
+			man_vel_sp(1) /= man_vel_hor_length;
 		}
 
 		/* reset position setpoint to current position if needed */
 		reset_pos_sp();
 	}
 
-	/* scale requested velocity setpoint to cruisespeed and rotate around yaw */
-	math::Vector<3> cruising_scale(_params.vel_cruise_xy,
-				       _params.vel_cruise_xy,
-				       (req_vel_sp_z > 0.0f) ? _params.vel_max_down : _params.vel_max_up);
-	math::Vector<3> req_vel_sp_scaled(req_vel_sp_xy(0), req_vel_sp_xy(1), req_vel_sp_z);
+	/* prepare yaw to rotate into NED frame */
+	float yaw_input_fame = _control_mode.flag_control_fixed_hdg_enabled ? _yaw_takeoff : _local_pos.yaw;
 
-	/* scale velocity setpoint to cruise speed (m/s) and rotate around yaw to NED frame */
-	math::Matrix<3, 3> R_input_fame;
+	/* prepare cruise speed (m/s) vector to scale the velocity setpoint */
+	matrix::Vector3f vel_cruise_scale(_params.vel_cruise_xy,
+					  _params.vel_cruise_xy,
+					  (man_vel_sp(2) > 0.0f) ? _params.vel_max_down : _params.vel_max_up);
 
-	if (_control_mode.flag_control_fixed_hdg_enabled) {
-		R_input_fame.from_euler(0.0f, 0.0f, _yaw_takeoff);
+	/* setpoint in NED frame and scaled to cruise velocity */
+	man_vel_sp = matrix::Dcmf(matrix::Eulerf(0.0f, 0.0f, yaw_input_fame)) * man_vel_sp.emult(vel_cruise_scale);
 
-	} else {
-		R_input_fame.from_euler(0.0f, 0.0f, _local_pos.yaw);
-
-	}
-
-	req_vel_sp_scaled = R_input_fame * req_vel_sp_scaled.emult(
-				    cruising_scale); // in NED and scaled to actual velocity;
 
 	/*
-	 * assisted velocity mode: user controls velocity, but if	velocity is small enough, position
+	 * assisted velocity mode: user controls velocity, but if velocity is small enough, position
 	 * hold is activated for the corresponding axis
 	 */
 
+	/* want to get/stay in altitude hold if user has z stick in the middle (accounted for deadzone already) */
+	const bool alt_hold_desired = _control_mode.flag_control_altitude_enabled && fabsf(man_vel_sp(2)) < FLT_EPSILON;
+
+	/* want to get/stay in position hold if user has xy stick in the middle (accounted for deadzone already) */
+	const bool pos_hold_desired = _control_mode.flag_control_position_enabled &&
+				      fabsf(man_vel_sp(0)) < FLT_EPSILON && fabsf(man_vel_sp(1)) < FLT_EPSILON;
 
```

Comment:
That's intentional

Because you used the exact same conditions in your smooth transition check and it's just not readable to have the same long expression over and over again.